In [21]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

### i imported libraries now i will read data


In [17]:
data = pd.read_csv(r'C:\Users\muaz\Desktop\Fraud.csv')

In [44]:
data.head(20)

step      type     amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
0      1   PAYMENT    9839.64  C1231006815      170136.00       160296.36   
1      1   PAYMENT    1864.28  C1666544295       21249.00        19384.72   
2      1  TRANSFER     181.00  C1305486145         181.00            0.00   
3      1  CASH_OUT     181.00   C840083671         181.00            0.00   
4      1   PAYMENT   11668.14  C2048537720       41554.00        29885.86   
5      1   PAYMENT    7817.71    C90045638       53860.00        46042.29   
6      1   PAYMENT    7107.77   C154988899      183195.00       176087.23   
7      1   PAYMENT    7861.64  C1912850431      176087.23       168225.59   
8      1   PAYMENT    4024.36  C1265012928        2671.00            0.00   
9      1     DEBIT    5337.77   C712410124       41720.00        36382.23   
10     1     DEBIT    9644.94  C1900366749        4465.00            0.00   
11     1   PAYMENT    3099.97   C249177573       20771.00        17671.03   
12     1   PAYMENT    2560.74  C1648232591        5070.00         2509.26   
13     1   PAYMENT   11633.76  C1716932897       10127.00            0.00   
14     1   PAYMENT    4098.78  C1026483832      503264.00       499165.22   
15     1  CASH_OUT  229133.94   C905080434       15325.00            0.00   
16     1   PAYMENT    1563.82   C761750706         450.00            0.00   
17     1   PAYMENT    1157.86  C1237762639       21156.00        19998.14   
18     1   PAYMENT     671.64  C2033524545       15123.00        14451.36   
19     1  TRANSFER  215310.30  C1670993182         705.00            0.00   

       nameDest  oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  
0   M1979787155             0.0            0.00        0               0  
1   M2044282225             0.0            0.00        0               0  
2    C553264065             0.0            0.00        1               0  
3     C38997010         21182.0            0.00        1               0  
4   M1230701703             0.0            0.00        0               0  
5    M573487274             0.0            0.00        0               0  
6    M408069119             0.0            0.00        0               0  
7    M633326333             0.0            0.00        0               0  
8   M1176932104             0.0            0.00        0               0  
9    C195600860         41898.0        40348.79        0               0  
10   C997608398         10845.0       157982.12        0               0  
11  M2096539129             0.0            0.00        0               0  
12   M972865270             0.0            0.00        0               0  
13   M801569151             0.0            0.00        0               0  
14  M1635378213             0.0            0.00        0               0  
15   C476402209          5083.0        51513.44        0               0  
16  M1731217984             0.0            0.00        0               0  
17  M1877062907             0.0            0.00        0               0  
18   M473053293             0.0            0.00        0               0  
19  C1100439041         22425.0            0.00        0               0

In [11]:
df = pd.DataFrame(data)

In [50]:
df.head()

step      type    amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
0     1   PAYMENT   9839.64  C1231006815       170136.0       160296.36   
1     1   PAYMENT   1864.28  C1666544295        21249.0        19384.72   
2     1  TRANSFER    181.00  C1305486145          181.0            0.00   
3     1  CASH_OUT    181.00   C840083671          181.0            0.00   
4     1   PAYMENT  11668.14  C2048537720        41554.0        29885.86   

      nameDest  oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  
0  M1979787155             0.0             0.0        0               0  
1  M2044282225             0.0             0.0        0               0  
2   C553264065             0.0             0.0        1               0  
3    C38997010         21182.0             0.0        1               0  
4  M1230701703             0.0             0.0        0               0

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6362620 entries, 0 to 6362619
Data columns (total 11 columns):
 #   Column          Dtype  
---  ------          -----  
 0   step            int64  
 1   type            object 
 2   amount          float64
 3   nameOrig        object 
 4   oldbalanceOrg   float64
 5   newbalanceOrig  float64
 6   nameDest        object 
 7   oldbalanceDest  float64
 8   newbalanceDest  float64
 9   isFraud         int64  
 10  isFlaggedFraud  int64  
dtypes: float64(5), int64(3), object(3)
memory usage: 534.0+ MB


### checking no of missing value in each column

In [28]:
df.isnull().sum()

step              0
type              0
amount            0
nameOrig          0
oldbalanceOrg     0
newbalanceOrig    0
nameDest          0
oldbalanceDest    0
newbalanceDest    0
isFraud           0
isFlaggedFraud    0
dtype: int64

### luckily we dont have any missing value

### distribution of legit transacition and fraud transaction

In [32]:
df['isFraud'].value_counts()

0    6354407
1       8213
Name: isFraud, dtype: int64

####  number of legit transaction = 6354407 while fraud transaction = 8213, there is huge gap in between normal and fraud trans. here 0 is normal while 1 is fraud

### seprating data for analysis

In [39]:
legit = df[df.isFraud == 0]
fraud = df[df.isFraud == 1]

In [41]:
print(legit.shape)
print(fraud.shape)

(6354407, 11)
(8213, 11)


### statistical measure of data

In [47]:
legit.amount.describe()

count    6.354407e+06
mean     1.781970e+05
std      5.962370e+05
min      1.000000e-02
25%      1.336840e+04
50%      7.468472e+04
75%      2.083648e+05
max      9.244552e+07
Name: amount, dtype: float64

In [45]:
fraud.amount.describe()

count    8.213000e+03
mean     1.467967e+06
std      2.404253e+06
min      0.000000e+00
25%      1.270913e+05
50%      4.414234e+05
75%      1.517771e+06
max      1.000000e+07
Name: amount, dtype: float64

In [54]:
df['amount'].max()


92445516.64

In [55]:
df['amount'].min()

0.0

### compare the value of both normal and fraud trans by groupby function

In [59]:
df.groupby('isFraud').mean()

step        amount  oldbalanceOrg  newbalanceOrig  \
isFraud                                                            
0        243.235663  1.781970e+05   8.328287e+05   855970.228109   
1        368.413856  1.467967e+06   1.649668e+06   192392.631836   

         oldbalanceDest  newbalanceDest  isFlaggedFraud  
isFraud                                                  
0          1.101421e+06    1.224926e+06        0.000000  
1          5.442496e+05    1.279708e+06        0.001948

## under - sampling

### build a sample dataset containing similar distribution of normal and fraud trans.

### here fraud trans is 8213 and normal is 6354407

In [60]:
legit_sample = legit.sample(n=8213)

In [61]:
legit_sample.count()

step              8213
type              8213
amount            8213
nameOrig          8213
oldbalanceOrg     8213
newbalanceOrig    8213
nameDest          8213
oldbalanceDest    8213
newbalanceDest    8213
isFraud           8213
isFlaggedFraud    8213
dtype: int64

### now legit_sample = fraud in no of trans.

### concatenating two data frame

In [64]:
new_df = pd.concat([legit_sample , fraud], axis =0)

In [67]:
new_df.tail()

step      type      amount     nameOrig  oldbalanceOrg  \
6362615   743  CASH_OUT   339682.13   C786484425      339682.13   
6362616   743  TRANSFER  6311409.28  C1529008245     6311409.28   
6362617   743  CASH_OUT  6311409.28  C1162922333     6311409.28   
6362618   743  TRANSFER   850002.52  C1685995037      850002.52   
6362619   743  CASH_OUT   850002.52  C1280323807      850002.52   

         newbalanceOrig     nameDest  oldbalanceDest  newbalanceDest  isFraud  \
6362615             0.0   C776919290            0.00       339682.13        1   
6362616             0.0  C1881841831            0.00            0.00        1   
6362617             0.0  C1365125890        68488.84      6379898.11        1   
6362618             0.0  C2080388513            0.00            0.00        1   
6362619             0.0   C873221189      6510099.11      7360101.63        1   

         isFlaggedFraud  
6362615               0  
6362616               0  
6362617               0  
6362618               0  
6362619               0

new_df.describe()

In [72]:
new_df['isFraud'].value_counts() 

0    8213
1    8213
Name: isFraud, dtype: int64

### now data is balanced

In [73]:
new_df.groupby('isFraud').mean()

step        amount  oldbalanceOrg  newbalanceOrig  \
isFraud                                                            
0        243.496530  1.862130e+05   8.187116e+05   842139.111818   
1        368.413856  1.467967e+06   1.649668e+06   192392.631836   

         oldbalanceDest  newbalanceDest  isFlaggedFraud  
isFraud                                                  
0          1.136684e+06    1.277039e+06        0.000000  
1          5.442496e+05    1.279708e+06        0.001948

### by looking into mean value and the other parameter which i saw by using describe function, column isFlaggedFraud , type ,nameOrig and nameDest Column which are  not important for analysis, this field will not affect model so i will delete it


In [75]:
new_df.drop('isFlaggedFraud', axis =1 , inplace = True)


In [90]:
new_df.drop('type', axis =1 , inplace = True)

In [106]:
new_df.drop('nameOrig', axis =1 , inplace = True)

KeyError: "['nameOrig'] not found in axis"

In [113]:
new_df.drop('nameDest', axis =1 , inplace = True)

### now spiliting data into features and target

In [114]:
X = new_df.drop(columns = 'isFraud', axis = 1)
Y = new_df['isFraud']

In [115]:
print(X)

         step      amount  oldbalanceOrg  newbalanceOrig  oldbalanceDest  \
5022206   353    26151.26     3739702.76      3765854.02      4031734.21   
5417823   378    21464.37        6074.00            0.00            0.00   
387059     17   253554.81       50540.00       304094.81         8539.00   
3756013   279   138048.95           0.00            0.00      1615928.51   
3987277   298   233472.24      103253.00            0.00       109245.87   
...       ...         ...            ...             ...             ...   
6362615   743   339682.13      339682.13            0.00            0.00   
6362616   743  6311409.28     6311409.28            0.00            0.00   
6362617   743  6311409.28     6311409.28            0.00        68488.84   
6362618   743   850002.52      850002.52            0.00            0.00   
6362619   743   850002.52      850002.52            0.00      6510099.11   

         newbalanceDest  
5022206      4005582.96  
5417823            0.00  
387059   

In [116]:
print(Y)

5022206    0
5417823    0
387059     0
3756013    0
3987277    0
          ..
6362615    1
6362616    1
6362617    1
6362618    1
6362619    1
Name: isFraud, Length: 16426, dtype: int64


### split data into training and testing data 

In [117]:
X_train , X_test , Y_train, Y_test = train_test_split(X, Y , test_size = 0.2 , stratify = Y , random_state = 2 )

#### take 20% of data as test data while remaining as train data

In [121]:
print(X.shape , X_train.shape , X_test.shape )

(16426, 6) (13140, 6) (3286, 6)


### model training by logistic  regression

In [119]:
model = LogisticRegression()

In [122]:
model.fit(X_train , Y_train)

LogisticRegression()

### Model Evaluation 

### Accuracy Score 

In [123]:
X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)

                        

In [124]:
print('Accuracy on Training data',training_data_accuracy )

Accuracy on Training data 0.9068493150684932


#### prediction percentage is 90% accurate

### now accuracy on test data 

In [125]:
X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)


In [126]:
print('Accuracy on Test data',test_data_accuracy )

Accuracy on Test data 0.9013998782714546


### test data accuracy is also about 90% 

###  here test and train data have around same accuracy wich means model works well it is neither overfitted or underfitted